### PR TITLE
SpecAugment after first SCF layer

### PR DESCRIPTION
**Pull request by Max-Ryujin**
Added the ability to perform unsorted specaug on the first layer

### CODE REVIEW

File: users/vieting/experiments/switchboard/hybrid/feat/baseline_args.py [hunk 1]
```diff
@@ -208,14 +208,19 @@ def get_returnn_config(
         network["source"] = {"class": "copy", "from": "features"}
     else:
         if specaug_mask_sorting:
+            assert specaug_after_first_layer, "Sorted specaug is only possible after the first layer!"
```
Comment (vieting) [inline users/vieting/experiments/switchboard/hybrid/feat/baseline_args.py:211]: `specaug_after_first_layer` is not defined, right?

File: users/vieting/experiments/switchboard/hybrid/feat/baseline_args.py [hunk 2]
```diff
@@ -208,14 +208,19 @@ def get_returnn_config(
...
-            network["source"] = specaug_layer_jingjing(in_layer=["features"])
+            if specaug_after_first_layer:
+                network["features"]["subnetwork"]["specaug"] = specaug_layer_jingjing(in_layer=["conv_h_act"])
+                network["features"]["subnetwork"]["conv_h_split"]["from"] = "specaug"
+                network["source"] = {"class": "copy", "from": "features"}
+            else:
+                network["source"] = specaug_layer_jingjing(in_layer=["features"])
 
         network = fix_network_for_sparse_output(network)
-
```
Comment (vieting) [inline users/vieting/experiments/switchboard/hybrid/feat/baseline_args.py:218]: I think this empty line can stay.

File: users/vieting/experiments/switchboard/hybrid/feat/experiments.py [hunk 3]
```diff
@@ -192,7 +192,11 @@ def run_baseline_scf():
     nn_args = get_nn_args_baseline(
         nn_base_args={
             "scf": dict(
-                returnn_args=dict(batch_size=14000),
```
Comment (vieting) [inline users/vieting/experiments/switchboard/hybrid/feat/experiments.py:195]: I did run this setup successfully with batch size 14k, so please do not change that. You can import this job from my setup.

File: users/vieting/experiments/switchboard/hybrid/feat/experiments.py [hunk 4]
```diff
@@ -192,7 +192,11 @@ def run_baseline_scf():
...
+                returnn_args=dict(batch_size=7000, extra_args=dict(accum_grad_multiple_step=2)),
+                feature_args={"class": "ScfNetwork", "size_tf": 256 // 2, "stride_tf": 10 // 2}
+            ),
+            "scf_first_layer": dict(
```
Comment (vieting) [inline users/vieting/experiments/switchboard/hybrid/feat/experiments.py:198]: The alias should reflect that first layer refers to SpecAugment.

File: users/vieting/experiments/switchboard/hybrid/feat/experiments.py [hunk 5]
```diff
@@ -192,7 +192,11 @@ def run_baseline_scf():
...
+                returnn_args=dict(batch_size=3500, specaug_after_first_layer=True, extra_args=dict(accum_grad_multiple_step=4)),
```
Comment (vieting) [inline users/vieting/experiments/switchboard/hybrid/feat/experiments.py:199]: Is it really necessary to split by 4? Is that for 11GB or 24GB GPUs?

**Review by vieting — CHANGES_REQUESTED** [review]

File: users/vieting/experiments/switchboard/hybrid/feat/experiments.py (hunk 5 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/hybrid/feat/experiments.py:199]: Yes it is necessary even for 24GB.

File: users/vieting/experiments/switchboard/hybrid/feat/baseline_args.py [hunk 6]
```diff
@@ -208,14 +209,20 @@ def get_returnn_config(
...
+
```
Comment (vieting) [inline users/vieting/experiments/switchboard/hybrid/feat/baseline_args.py:225]:
```suggestion

```

File: users/vieting/experiments/switchboard/hybrid/feat/baseline_args.py [hunk 7]
```diff
@@ -208,14 +209,19 @@ def get_returnn_config(
...
         else:
-            network["source"] = specaug_layer_jingjing(in_layer=["features"])
+            if specaug_after_first_layer:
+                network["features"]["subnetwork"]["specaug"] = specaug_layer_jingjing(in_layer=["conv_h_act"])
+                network["features"]["subnetwork"]["conv_h_split"]["from"] = "specaug"
+                network["source"] = {"class": "copy", "from": "features"}
+            else:
+                network["source"] = specaug_layer_jingjing(in_layer=["features"])
 
         network = fix_network_for_sparse_output(network)
```
Comment (vieting) [inline users/vieting/experiments/switchboard/hybrid/feat/baseline_args.py:224]:
```suggestion
        network = fix_network_for_sparse_output(network)

```

**Pull request merged**